### ISSUE TITLE
Tensors must be CUDA and dense

### ISSUE THREAD

**Issue by yjw1029** [huggingface/trl#715]
Get error when using PPOTrainer with multi GPUs:
```
File "/home/t-jingweiyi/miniconda3/envs/torch2/lib/python3.10/site-packages/trl/trainer/ppo_trainer.py", line 879, in gather_stats
stats = self.gather_stats(stats)
....
RuntimeError: RuntimeError Tensors must be CUDA and dense: 
```
My script is similar to https://github.com/huggingface/trl/blob/main/examples/research_projects/stack_llama/scripts/rl_training.py

The reason might be `ppo/mean_scores` and `ppo/std_scores` are not CUDA.

**Comment by BillChan226** [comment]
I have also encountered this. This is frequently encountered when you try to use multi-gpus. You can trace back to where the last error is raised and print the tensor to see where it is. For me, it happens on line 864 of ppo_trainer.py:
`        for k, v in stats.items():
            if isinstance(v, torch.Tensor):
                if v.is_cuda == False:
                    v = v.to(self.current_device)
                dist.all_reduce(v, dist.ReduceOp.SUM)
                v /= self.accelerator.num_processes
            stats[k] = v
        return stats`
I print `v`, and it turns out it is on cpu. So I drop it to `self.current_device` if it is not on CUDA. This resolves my problem.

**Comment by BillChan226** [comment]
You can add my contact and we can discuss how to fine-tune PPO with RLHF. I'm also encountering several issues. I'd love to discuss with someone.

wechat: czr1158764709 
email: chen4399@purdue.edu

**Comment by lvwerra** [comment]
@younesbelkada I think for these kind of things it would also be great to have GPU tests.

**Comment by ywangmy** [comment]
> I have also encountered this. This is frequently encountered when you try to use multi-gpus. You can trace back to where the last error is raised and print the tensor to see where it is. For me, it happens on line 864 of ppo_trainer.py: ` for k, v in stats.items(): if isinstance(v, torch.Tensor): if v.is_cuda == False: v = v.to(self.current_device) dist.all_reduce(v, dist.ReduceOp.SUM) v /= self.accelerator.num_processes stats[k] = v return stats` I print `v`, and it turns out it is on cpu. So I drop it to `self.current_device` if it is not on CUDA. This resolves my problem.

Which `v` is on cpu?

**Comment by BillChan226** [comment]
> Which v is on cpu?

The value `v`, of dict `k, v in stats.items()`

**Comment by ywangmy** [comment]
> > Which v is on cpu?
> 
> The value `v`, of dict `k, v in stats.items()`

I mean, which `k` corresponds to the `v` on cpu?

**Comment by cloneofsimo** [comment]
I have encountered this as well, looks like stats is numpy array, which is not directly converted to torch tensor.
For me, I just changed the code to:

```python
    # Gather/Reduce stats from all processes
        if self.is_distributed:
            if self.accelerator.is_main_process:
                stats = stats
```

**Comment by ywangmy** [comment]
A similar issue #451 has been resolved in #513. Installing from source works for me.

**Comment by younesbelkada** [comment]
Awesome! It seems installing from source fixes the issue, feel free to re-open if that's not the case
